Release Candidate: Refactor + Selective Authentication + New Configuration

thomseddon opened the pull request:
This is a release candidateof a refactor we've been working on, this was primarily motivated by the desire to selectively apply authentication (e.g. to some paths only).

Tasks

- [x] Refactor codebase into smaller chunks
- [x] Support selective authentication
- [x] Improve logging
- [x] Facilitate multiple authentication providers
- [x] ~~Use goth for authentication~~ (planned for 2.1)
- [x] Support multiple config files
- [x] Maintain backwards compatibility for all flags
- [x] Update tests to use testify
- [x] Documentation


Full documentation: https://github.com/thomseddon/traefik-forward-auth/blob/dev/README.md

## Selective Authentication

Example config:
```
default-action = auth
secret = 123

providers.google.client-secret = 123
providers.google.client-id = 234

rules.1.action = allow
rules.1.rule = Host(`example.com`) && PathPrefix(`/api/public`) && Headers(`Content-Type`, `text/json`)

rules.two.action = allow
rules.two.rule = PathPrefix(`/public`)
```

The rule syntax is `rule.<name>.<param>`:

- name can be any string
- param should be one of:
  - action: `allow` to bypass authentication or `auth` to enforce authentication
  - rule: rule that will be parsed by traefik's v2 parser, documentation here: https://docs.traefik.io/v2.0/routing/routers/#rule

In the above example, the first rule means that requests to the host "example.com", with the path prefix "/api/public" and the "text/json" content type would be allowed without any OAuth authentication.

rdehuyss commented:
Any idea when this will be merged to master?

thomseddon commented:
Hopefully within the next couple of weeks :)

thomseddon commented:
Quick update - I've noticed that with many of the issues created, people seem to be using label based config, this makes me realise why traefik use the string based configuration as opposed to the toml structures....as such I will change this to support the same string format as traefik.

Updates soon :)

FuNK3Y commented:
Looking forward to that!

Once it is there I'll try to build the microsoft provider.

Will it still be possible to provide configuration with arguments / environnement variables after the refactor or will having a config file be mandatory?

thomseddon commented:
Yep, we'll support arguments and environment variables for sure, I can see that this how many prefer to do their config :)

andig commented:
@thomseddon not sure if this is obvious but it would be great if configuration could be done using docker labels like traefik does in addition to using the toml file. Really looking forward to use the new features in action!

jeliasson commented:
Looking forward to this merge! 
Anyone have plans to add [Azure AD](https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-protocols-oauth-code) as a provider?

thomseddon commented:
Sorry for the delays on this! I've been doing some bits on this for the last few days, should have decent progress to push tomorrow!

Yes, I would like to support other providers, this will only include Google but then I'd like to look at that feature next...

andig commented:
> Yes, I would like to support other providers, this will only include Google but then I'd like to look at that feature next...

Would love github, too in order to not give google all my data...

FuNK3Y commented:
@thomseddon could leveraging lib like https://github.com/markbates/goth as provider make sense?

thomseddon commented:
I've pushed up our recent progress and edited the description, things to note:

- We've changed the rule syntax to use Traefik's version 2 string parser directly (see notes above)
- Code structure changed

We still have a few more things we need to box off (mainly maintaining full backwards compat) then we should be good to go!

thomseddon commented:
OK, I've implemented the third and final iteration of the rule syntax:

```
rules.1.action = allow
rules.1.rule = Host(`example.com`) && PathPrefix(`/api/public`) && Headers(`Content-Type`, `text/json`)

rules.two.action = allow
rules.two.rule = PathPrefix(`/public`)
```

I've updated the original PR comment with full details.

This is now more or less finished, next steps will be to fully implement config backwards compatibility and to finish writing tests to check the implementation. I have tested using Goth for authentication (see expr/goth branch), this should work fine but I'm going to put that in the next feature release.

Feedback on the latest rule configuration or the general state of the PR is appreciated as always...

thomseddon commented:
Looking really good now and getting close to a release! I've added a lot more tests and made a few fixes.

We're now completely backwards compatible (this was important as original examples used "latest" docker image) and well tested.

We'll be deploying this into production next week (:crossed_fingers: :laughing:) if all goes well the only thing outstanding is updating the docs.

andig commented:
@thomseddon seems you‘ve removed the docker event listener I had proposed- is this intentional?

thomseddon commented:
Sorry, I was cleaning up this branch in preparation for merging which has had some ugly repercussions on your PR. I haven't merged your branch yet, I'm planning on doing that in 2.1 :+1:

andig commented:
One thing I've noticed reading the traefik docs for 2.0 is that the selective authentication feature introduced here is probably becoming obsolete. With traefik 2, forward auth becomes a middleware and middlewares can selectively be enabled per frontend (which determines routing). 

I'm wondering if its worth adding this now for benefit of traefik 1 or if we shouldn't target traefik 2 right away?

thomseddon commented:
Hmm, my use case was actually to bypass authentication for specific paths, I'm not sure it would be possible to achieve this with traefik middleware? But please let me know if I've missed something.
As I've just used the traefik rules library it really does add much complexity, I was quite happy with how it turned out.

As an aside, I do wonder if the best endgame for this service would actually be to become an official/core traefik "auth-oauth" middleware?

andig commented:
> Hmm, my use case was actually to bypass authentication for specific paths, I'm not sure it would be possible to achieve this with traefik middleware?

I'm only catching up, but I think that's exactly what you can do. Consider this example:

    labels:
    # frontend
    - "traefik.http.routers.frontend.rule=Host(`my.io`)"
    - "traefik.http.routers.frontend.service=frontend"
    - "traefik.http.services.frontend.loadbalancer.server.port=8080"
    # api
    - "traefik.http.routers.api.rule=Host(`my.io`) && PathPrefix(`/api`)"
    - "traefik.http.routers.api.middlewares=api"
    - "traefik.http.routers.api.service=api"
    - "traefik.http.middlewares.api.stripprefix.prefixes=/api"
    - "traefik.http.middlewares.api.compress=true"
    - "traefik.http.services.api.loadbalancer.server.port=8080"

I'm routing different paths to different ports by using the `stripprefix` middleware for one case. I believe the same way you could use the `forwardauth` middleware selectively. Just realized this when working on the above setup.

> As an aside, I do wonder if the best endgame for this service would actually be to become an official/core traefik "auth-oauth" middleware?

Looking at https://docs.traefik.io/v2.0/middlewares/overview/ it seems that `ForwardAuth` is the approach for doing this- no need for OAuth in the core?

thomseddon commented:
I'll do some testing, if they'll support multiple services/routers to be defined in labels then we're onto a winner!

It does seem Forward Auth is the preferred path, but there are a few long standing issues that have been kept open on traefik regarding authentication that makes me think they may one day want to grow middleware support to something like this - who knows, I guess it doesn't much matter either way!

thomseddon commented:
I've overhauled the tests to use testify, made quite a few fixes and updated all the documentation :)

This is now released as "v2.0.0-rc1" on docker hub, we have deployed this into production yesterday and it's working well! I would hugely appreciate if anyone else would try deploying this and letting me know how it goes!

It should just be a case of updating your docker-compose.yml:
```
 traefik-forward-auth:
    image: thomseddon/traefik-forward-auth:v2.0.0-rc1
    environment:
      - CLIENT_ID=your-client-id
      - CLIENT_SECRET=your-client-secret
      - SECRET=something-random
```

dalanmiller commented:
@thomseddon can you provide an example of limiting auth per subdomain with your 2.0? I'm a bit confused with the above using the new Traefik 2.0 config.

thomseddon commented:
@dalanmiller yes sure, are the subdomains routed to the same or different services?

Please feel free to post an example/extract of your docker-compose.yml :)

dalanmiller commented:
@thomseddon the set up I have is basically a bunch of HTTP services that should be behind the forward auth (e.g. Nextcloud) and then some other services that are also HTTP but have to either be exposed (Gitalb) or go behind basic auth. [TTRSS](https://tt-rss.org/) is a good example of having to be exposed or basic auth as the native mobile app only supports basic auth or I think cert auth 🤔 .

But anyway, would be curious to see how this is done with the new version! Thank you so much for your help!

lbischof commented:
I can't seem to get the rules working:
```
client-id = xxx
client-secret = xxx
secret = xxx
cookie-domains=domain.ch
auth-host=auth.domain.ch
whitelist=xx1@gmail.com,xx2@gmail.com

rule.1.action = allow
rule.1.rule = Host(`webdav.domain.ch`)
```
But when I access `webdav.domain.ch` it still redirects to `accounts.google.com`.

FYI: I think the config parsing fails if the rules are at the top.

thomseddon commented:
@dalanmiller at the moment, the way to do this would be to define rules to allow them:
```
rule.subdomain1.action = allow
rule.subdomain1.rule = Host(`subdomain1.example.com`)
```
You could combine a few into one rule (`Host(...) || Host(...)`) for brevity. For the 2.1 release we will be create a docker integration so we can use labels, similar to how traefik config is defined.

@lbischof I have just deployed some host rules and I can see the issue (`X-Forwarded-Host` is not being copied into the request for matching) I will be working on a fix for that. As a bit of help, I added some extra debug logs today that show which rule a given request is matching, this helped me!

thomseddon commented:
So, the Host matching wasn't working as the original request had to be copied through to the matcher, but it also turns our traefik have their own special host matcher which was not compatible, so that required a little work around.

In addition, rules were not being parsed from config files! This required changes to the upstream go-flags library (https://github.com/jessevdk/go-flags/pull/304) but it's now working with extra tests!

dalanmiller commented:
```
$ docker-compose -f docker-compose.yml up --force-recreate traefik-forward-auth
Recreating docker_traefik-forward-auth_1 ... done
Attaching to docker_traefik-forward-auth_1
traefik-forward-auth_1  | time="2019-05-11T00:18:37Z" level=fatal msg="google.providers.client-id, google.providers.client-secret must be set"
docker_traefik-forward-auth_1 exited with code 1
```

Getting the above error when I definitely have the following in the env vars for `traefik-forward-auth` service. 

```
traefik-forward-auth:
    image: thomseddon/traefik-forward-auth:v2.0.0-rc1
    command: ./traefik-forward-auth --rule.1.action=allow --rule.1.rule="Host(`host.example.com`)"
    environment:
      - PROVIDERS_GOOGLE_CLIENT_ID=FOO123
      - PROVIDERS_GOOGLE_CLIENT_SECRET=BAR123
```

Is something happening when `PROVIDERS_GOOGLE_CLIENT_SECRET` is being parsed? I tried changing the last underscore to a hyphen but got the same result.

thomseddon commented:
@dalanmiller that wasn't working in rc1, if you try with the development tag then that should work (as well as rules passed via config file

dalanmiller commented:
Will give it a try!

lbischof commented:
I am building my own container for arm and receive an error during the `go get` command:
```
go: github.com/thomseddon/go-flags@v1.4.1-0.20190507181358-ce437f05b7fb: unknown revision ce437f05b7fb
```
I replaced the line in `go.mod` with:
```
github.com/thomseddon/go-flags v1.4.1-0.20190507184247-a3629c504486
```

thomseddon commented:
@lbischof ah yes, I modified that commit in my go-flags fork (hopefully some derivative will be merged into go-flags soon)
Thanks for pointing that out, that has saved the build failing on the first release 🙈

I'm going to modify/add some logging ok n the branch (no functional changes) and aim to release next week.

Thank you everyone for you feedback so far, if anyone else on this thread is willing to test this branch then please let me know the outcome!

dalanmiller commented:
@thomseddon looks like the `development` tag still isn't working and going to `v2.0.0-rc2` results in a runtime error. Let me know if I can provide any information to help debug.

thomseddon commented:
@dalanmiller I've just tested this locally and with `PROVIDERS_GOOGLE_CLIENT_ID=1 PROVIDERS_GOOGLE_CLIENT_SECRET=1` and it seems to work OK, I did actually add a test for this during development: https://github.com/thomseddon/traefik-forward-auth/blob/dev/internal/config_test.go#L196-L201

Perhaps you're not on the latest development tag, I've pushed a new tag `2.0.0-rc3`, can you test with that?

(I also noted the error you got is formatted incorrectly, I've fixed that in my local branch and will commit along with some other logging changes I'm quickly nipping in prior to release)

seuf commented on internal/auth.go:
```diff
@@ -0,0 +1,343 @@
+
+// // Return url
+func returnUrl(r *http.Request) string {
+	path := r.Header.Get("X-Forwarded-Uri")
```
Maybe you should return `X-Forwarded-Prefix` do directly redirect to the initial request.

thomseddon replied:
Sorry I'm not quite sure I follow, in https://docs.traefik.io/basics/ it says `X-Forwarded-Prefix` header will be "the stripped prefix path" following a PathPrefixStrip or PathPrefixStripRegex match, so I don't think using this will not be correct in all circumstances.

*However*, using those two matchers can cause issues and there is a fix for this in #39 which I will merge into this branch :)

seuf replied:
OK, cool. Any eta for the next release ?
Thanks for your good job ! Love it !
